File: backend/app/schemas/cafeteria.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# Categorías 
class CategoriaResponse(BaseModel):
    """Schema de respuesta para una categoría de cafetería."""
    id: uuid.UUID
    nombre: str
    descripcion: str | None = None
    imagen_url: str | None = None
    orden: int
    activa: bool
    productos: list["ProductoResponse"] = []

    class Config:
        from_attributes = True


class CategoriaCreate(BaseModel):
    """Schema para crear una categoría."""
    nombre: str = Field(..., min_length=1, max_length=100)
    descripcion: str | None = None
    imagen_url: str | None = None
    orden: int = 0


class CategoriaUpdate(BaseModel):
    """Schema para actualizar una categoría."""
    nombre: str | None = Field(None, min_length=1, max_length=100)
    descripcion: str | None = None
    imagen_url: str | None = None
    orden: int | None = None
    activa: bool | None = None


# Productos

class ProductoResponse(BaseModel):
    """Schema de respuesta para un producto de cafetería."""
    id: uuid.UUID
    categoria_id: uuid.UUID
    nombre: str
    descripcion: str | None = None
    imagen_url: str | None = None
    precio: Decimal
    disponible: bool
    destacado: bool
    orden: int

    class Config:
        from_attributes = True


class ProductoCreate(BaseModel):
    """Schema para crear un producto."""
    categoria_id: uuid.UUID
    nombre: str = Field(..., min_length=1, max_length=150)
    descripcion: str | None = None
    imagen_url: str | None = None
    precio: Decimal = Field(..., ge=0)
    disponible: bool = True
    destacado: bool = False
    orden: int = 0


class ProductoUpdate(BaseModel):
    """Schema para actualizar un producto."""
    categoria_id: uuid.UUID | None = None
    nombre: str | None = Field(None, min_length=1, max_length=150)
    descripcion: str | None = None
    imagen_url: str | None = None
    precio: Decimal | None = Field(None, ge=0)
    disponible: bool | None = None
    destacado: bool | None = None
    orden: int | None = None

File: frontend/lib/models/cafeteria.dart

/// Modelos de datos Cafetería.
///
/// Define la estructura del menú de la cafetería, organizando los artículos
/// en categorías y gestionando precios, disponibilidad y destacados.
library;

import 'package:reservives/config/constants.dart';

/// Representa una sección del menú
class CategoriaCafeteria {
  final String id;
  final String nombre;
  final String? descripcion;
  final String? imagenUrl;
  final int orden;
  final bool activa;
  final List<ProductoCafeteria> productos;

  const CategoriaCafeteria({
    required this.id,
    required this.nombre,
    this.descripcion,
    this.imagenUrl,
    required this.orden,
    required this.activa,
    required this.productos,
  });

  /// Crea una categoría y mapea recursivamente su lista de productos desde JSON.
  factory CategoriaCafeteria.fromJson(Map<String, dynamic> json) {
    final productosList = json['productos'] as List? ?? [];
    return CategoriaCafeteria(
      id: json['id'] as String,
      nombre: json['nombre'] as String,
      descripcion: json['descripcion'] as String?,
      imagenUrl: AppConstants.resolveApiUrl(json['imagen_url'] as String?),
      orden: json['orden'] as int,
      activa: json['activa'] as bool,
      productos: productosList
          .map((p) => ProductoCafeteria.fromJson(p as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// Representa un artículo individual a la venta en la cafetería.
class ProductoCafeteria {
  final String id;
  final String categoriaId;
  final String nombre;
  final String? descripcion;
  final String? imagenUrl;
  final double precio;
  final bool disponible;
  final bool destacado;

  const ProductoCafeteria({
    required this.id,
    required this.categoriaId,
    required this.nombre,
    this.descripcion,
    this.imagenUrl,
    required this.precio,
    required this.disponible,
    required this.destacado,
  });

  factory ProductoCafeteria.fromJson(Map<String, dynamic> json) {
    final precioRaw = json['precio'];
    final precio = precioRaw is num
        ? precioRaw.toDouble()
        : double.tryParse(precioRaw.toString()) ?? 0.0;

    return ProductoCafeteria(
      id: json['id'] as String,
      categoriaId: json['categoria_id'] as String,
      nombre: json['nombre'] as String,
      descripcion: json['descripcion'] as String?,
      imagenUrl: AppConstants.resolveApiUrl(json['imagen_url'] as String?),
      precio: precio,
      disponible: json['disponible'] as bool,
      destacado: json['destacado'] as bool,
    );
  }
}
